src/libraries/System.Private.CoreLib/src/System/SpanHelpers.cs

```diff
@@ -403,5 +406,200 @@ public static unsafe void ClearWithReferences(ref IntPtr ip, nuint pointerSizeLe
+                    //     +-------------------------------+
+                    //     | A | B | C | D | E | F | G | H |
+                    //     +-------------------------------+
+                    //         --->
+                    //     +-------------------------------+
+                    //     | H | G | F | E | D | C | B | A |
+                    //     +-------------------------------+
+                    tempFirst = Avx2.PermuteVar8x32(tempFirst, reverseMask);
+                    tempLast = Avx2.PermuteVar8x32(tempLast, reverseMask);
+
+                    // Store the values into final location
+                    tempLast.StoreUnsafe(ref buf, firstOffset);
+                    tempFirst.StoreUnsafe(ref buf, lastOffset);
+                }
+                buf = ref Unsafe.Add(ref buf, numIters * numElements);
+                length -= numIters * numElements * 2;
+            }
+            else if (Sse2.IsSupported && (nuint)Vector128<int>.Count * 2 <= length)
+            {
+                nuint numElements = (nuint)Vector128<int>.Count;
+                nuint numIters = (length / numElements) / 2;
+                for (nuint i = 0; i < numIters; i++)
+                {
+                    nuint firstOffset = i * numElements;
+                    nuint lastOffset = length - ((1 + i) * numElements);
+
+                    // Load the values into vectors
+                    Vector128<int> tempFirst = Vector128.LoadUnsafe(ref buf, firstOffset);
+                    Vector128<int> tempLast = Vector128.LoadUnsafe(ref buf, lastOffset);
+
+                    // Shuffle to reverse each vector:
+                    //     +---------------+
+                    //     | A | B | C | D |
+                    //     +---------------+
+                    //          --->
+                    //     +---------------+
+                    //     | D | C | B | A |
+                    //     +---------------+
+                    tempFirst = Sse2.Shuffle(tempFirst, 0b00_01_10_11);
+                    tempLast = Sse2.Shuffle(tempLast, 0b00_01_10_11);
+
+                    // Store the values into final location
+                    tempLast.StoreUnsafe(ref buf, firstOffset);
+                    tempFirst.StoreUnsafe(ref buf, lastOffset);
+                }
+                buf = ref Unsafe.Add(ref buf, numIters * numElements);
+                length -= numIters * numElements * 2;
+            }
+
+            // Store any remaining values one-by-one
+            for (nuint i = 0; i < (length / 2); i++)
+            {
+                ref int firstInt = ref Unsafe.Add(ref buf, i);
+                ref int lastInt = ref Unsafe.Add(ref buf, length - 1 - i);
+                (lastInt, firstInt) = (firstInt, lastInt);
+            }
+        }
+
+        public static void Reverse(ref long buf, nuint length)
+        {
+            if (Avx2.IsSupported && (nuint)Vector256<long>.Count * 2 <= length)
```

Comment:
Can this be implemented using the cross-architecture Vector256/Vector128 intrinsics so that the optimization works on all architectures?

Reply:
This optimization relies on permute and shuffle operations, which don't yet have a cross-platform implementation for Vector256/Vector128 intrinsics ([Tanner's comment](https://github.com/dotnet/runtime/pull/64412#discussion_r801082171)). Once those are ready, this would be a good area to use them instead of the platform-specific intrinsics so all platforms benefit, but for now the PR is only for platforms that support AVX2/SSE2.